src/dfm/cleaning/deduper.py

```diff
@@ -351,65 +282,116 @@ def deduplicate(
+            # Initialise the multiprocessing
+            with Parallel(n_jobs=self.n_jobs) as parallel:
+                fn = delayed(
+                    partial(
+                        get_minhash,
+                        normalization_func=self.normalization_func,
+                        split_method=self.split_method,
+                        ngram_size=self.ngram_size,
+                        ngram_stride=self.ngram_stride,
+                        num_minhashes=self.num_minhashes,
+                        random_seed=self.random_seed,
+                    )
+                )
 
-                # Update the progress bar
-                pbar.update(len(batch))
-                pct_duplicated = 100 * duplicates / (1 + max_doc_idx)
-                desc = f"Deduplicating - {pct_duplicated:.2f}% near-duplicates found"
-                pbar.set_description(desc)
+                # Iterate over the batches
+                for batch in pbar:
+
+                    # Create a copy of the batch to ensure that we're not
+                    # modifying the original
+                    batch, batch_copy = it.tee(batch)
+
+                    # Compute size of the batch
+                    new_num_processed = num_processed + self.batch_size
+                    new_num_processed = min(new_num_processed, num_docs)
+                    batch_size = new_num_processed - num_processed
+
+                    # Define parameters used in batch progress bars
+                    pbar_params = dict(
+                        total=batch_size, leave=False, disable=(not self.verbose)
+                    )
+
+                    # Compute the fingerprint for the document
+                    pbar_params["desc"] = "Computing minhashes"
+                    with tqdm(batch, **pbar_params) as batch_pbar:
+                        minhashes = parallel(fn(doc) for _, doc in batch_pbar)
+
+                    # Iterate over the minhashes
+                    pbar_params["desc"] = "Deduplicating batch"
+                    with tqdm(batch_copy, **pbar_params) as batch_pbar:
+                        for (idx, doc), minhash in zip(batch_pbar, minhashes):
+
+                            # If the document is not a near-duplicate candidate
+                            # then store in the LSH cache and append it to the
+                            # JSONL output file
+                            candidates = self.lsh_cache.query(minhash)
+                            if len(candidates) == 0:
+
+                                # Insert the document into the LSH cache
+                                self.lsh_cache.insert(idx, minhash)
+
+                                # Store the non-duplicate document in the JSONL
+                                # output
+                                self._store_document(
+                                    id=idx, text=doc, output_path=output_path
+                                )
```

Comment:
Regarding what we talked about. If you don't want the function to write the non-duplicate you could simply yield the boolean here and then create a function `deduplicate_to_disk` which simply zips the boolean indicator.

This would also have the added benefit of splitting "output_dir" of the deduper (mask a and cache) with the output_dir of the deduped files

Reply:
@KennethEnevoldsen If I understand this correctly, you want it to compute the minhashes in parallel for the batch, and then yield one document at a time afterwards?

Also, if we want to split up the output directory, the mask and the deduplicated corpus should be together, as they are both corpus-specific. The only deduper information independent of corpus are the LSH cache and the config. I could save these in separate files - is that what you're eluding to here?